agent/taskresource/cgroup.go

```diff
@@ -0,0 +1,225 @@
+	"github.com/aws/amazon-ecs-agent/agent/resources/cgroup"
+	"github.com/aws/amazon-ecs-agent/agent/utils/ioutilwrapper"
+	"github.com/cihub/seelog"
+	"github.com/containerd/cgroups"
+	"github.com/pkg/errors"
+)
+
+const (
+	memorySubsystem         = "/memory"
+	memoryUseHierarchy      = "memory.use_hierarchy"
+	rootReadOnlyPermissions = os.FileMode(400)
+)
+
+var (
+	enableMemoryHierarchy = []byte(strconv.Itoa(1))
+)
+
+// CgroupResource represents Cgroup resource
+type CgroupResource struct {
+	control             cgroup.Control
+	CgroupRoot          string
+	CgroupMountPath     string
+	ioutil              ioutilwrapper.IOUtil
+	createdAt           time.Time
+	desiredStatusUnsafe CgroupStatus
+	knownStatusUnsafe   CgroupStatus
+	// lock is used for fields that are accessed and updated concurrently
+	lock sync.RWMutex
+}
+
+// NewCgroupResource is used to return an object that implements the Resource interface
+func NewCgroupResource(cgroupRoot string, cgroupMountPath string) *CgroupResource {
+	return &CgroupResource{
+		// control creates interface to cgroups library methods
+		control:         cgroup.New(),
+		CgroupRoot:      cgroupRoot,
+		CgroupMountPath: cgroupMountPath,
+		ioutil:          ioutilwrapper.NewIOUtil(),
+	}
+}
+
+// SetDesiredStatus safely sets the desired status of the resource
+func (c *CgroupResource) SetDesiredStatus(status CgroupStatus) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	c.desiredStatusUnsafe = status
+}
+
+// GetDesiredStatus safely returns the desired status of the task
+func (c *CgroupResource) GetDesiredStatus() CgroupStatus {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+
+	return c.desiredStatusUnsafe
+}
+
+// SetKnownStatus safely sets the currently known status of the resource
+func (c *CgroupResource) SetKnownStatus(status CgroupStatus) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	c.knownStatusUnsafe = status
+}
+
+// GetKnownStatus safely returns the currently known status of the task
+func (c *CgroupResource) GetKnownStatus() CgroupStatus {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+
+	return c.knownStatusUnsafe
+}
+
+// SetCreatedAt sets the timestamp for resource's creation time
+func (c *CgroupResource) SetCreatedAt(createdAt time.Time) {
+	if createdAt.IsZero() {
+		return
+	}
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	c.createdAt = createdAt
+}
+
+// GetCreatedAt sets the timestamp for resource's creation time
+func (c *CgroupResource) GetCreatedAt() time.Time {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+
+	return c.createdAt
+}
+
+func (c *CgroupResource) Create(task *api.Task) error {
+	err := c.cgroupInit()
+	if err != nil {
+		seelog.Criticalf("unable to setup '/ecs' cgroup: %v", err)
+		return err
+	}
+	err = c.setupTaskCgroup(task)
+	if err != nil {
+		seelog.Criticalf("unable to setup platform resources: %v", err)
+		return err
+	}
+	return nil
+}
+
+func (c *CgroupResource) cgroupInit() error {
+	if c.control.Exists(config.DefaultTaskCgroupPrefix) {
```

Comment:
The cgroup prefix route will be created at the agent level, right?  Which means it's only created once when the agent restart, do we still need this?

Reply:
As per our offline discussion, cgroup controller can be created on agent level and passed to `CgroupResource` object, and that can be used to create the cgroup prefix in agent as well.